embassy/src/channel/pubsub.rs

```diff
@@ -0,0 +1,613 @@
+    const fn new() -> Self {
+        const WAKER_INIT: Option<WakerRegistration> = None;
+        Self {
+            queue: Deque::new(),
+            next_message_id: 0,
+            subscriber_wakers: [WAKER_INIT; SUBS],
+            publisher_wakers: [WAKER_INIT; PUBS],
+        }
+    }
+}
+
+/// A subscriber to a channel
+///
+/// This instance carries a reference to the channel, but uses a trait object for it so that the channel's
+/// generics are erased on this subscriber
+pub struct Subscriber<'a, T: Clone> {
+    /// Our index into the channel
+    subscriber_index: usize,
+    /// The message id of the next message we are yet to receive
+    next_message_id: u64,
+    /// The channel we are a subscriber to
+    channel: &'a dyn PubSubBehavior<T>,
+}
+
+impl<'a, T: Clone> Subscriber<'a, T> {
+    /// Wait for a published message
+    pub fn next_message<'s>(&'s mut self) -> SubscriberWaitFuture<'s, 'a, T> {
+        SubscriberWaitFuture { subscriber: self }
+    }
+
+    /// Wait for a published message (ignoring lag results)
+    pub async fn next_message_pure(&mut self) -> T {
+        loop {
+            match self.next_message().await {
+                WaitResult::Lagged(_) => continue,
+                WaitResult::Message(message) => break message,
+            }
+        }
+    }
+
+    /// Try to see if there's a published message we haven't received yet.
+    ///
+    /// This function does not peek. The message is received if there is one.
+    pub fn try_next_message(&mut self) -> Option<WaitResult<T>> {
+        match self.channel.get_message(self.next_message_id) {
+            Some(WaitResult::Lagged(amount)) => {
+                self.next_message_id += amount;
+                Some(WaitResult::Lagged(amount))
+            }
+            result => {
+                self.next_message_id += 1;
+                result
+            }
+        }
+    }
+
+    /// Try to see if there's a published message we haven't received yet (ignoring lag results).
+    ///
+    /// This function does not peek. The message is received if there is one.
+    pub fn try_next_message_pure(&mut self) -> Option<T> {
+        loop {
+            match self.try_next_message() {
+                Some(WaitResult::Lagged(_)) => continue,
+                Some(WaitResult::Message(message)) => break Some(message),
+                None => break None,
+            }
+        }
+    }
+}
+
+impl<'a, T: Clone> Drop for Subscriber<'a, T> {
+    fn drop(&mut self) {
+        self.channel
+            .unregister_subscriber(self.subscriber_index, self.next_message_id)
+    }
+}
+
+/// Warning: The stream implementation ignores lag results and returns all messages.
+/// This might miss some messages without you knowing it.
+impl<'a, T: Clone> futures::Stream for Subscriber<'a, T> {
+    type Item = T;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = unsafe { self.get_unchecked_mut() };
```

Comment:
unsafe is not needed if Subscriber is Unpin

Reply:
Oh cool!